### ISSUE TITLE
How to cache multiple requests to the same query

### ISSUE THREAD

**Issue by shovel-kun** [ConrabOpto/mst-query#74]
Hi, 

I was trying to test and see if caching works however, `mst-query` seems to only cache the last request for a itemQuery, and discard previous requests made with different ids.

```js

test('useQuery - test multiple reactive request', async () => {
    const { render, q, queryClient } = setup();

    configureMobx({ enforceActions: 'never' });

    let id = observable.box('test');
    const Comp = observer(() => {
        const { query } = useQuery(q.itemQuery, {
            request: { id: id.get() },
            staleTime: 0,
        });
        return <div></div>;
    });
    render(<Comp />);

    await wait(0);
    expect(q.itemQuery.data?.id).toBe('test');

    id.set('different-test');
    await wait(0);
    expect(q.itemQuery.data?.id).toBe('different-test');
    expect(q.itemQuery.variables.request?.id).toBe('different-test');

    const queries = queryClient.queryStore.getQueries(ItemQuery, (query) => query);

    expect(queries.length).toBe(2);

    configureMobx({ enforceActions: 'observed' });
});

test('findAll - test multiple queries', () => {
    const { q, queryClient } = setup();

    q.itemQuery.query({ request: { id: 'test1' } });
    q.itemQuery.query({ request: { id: 'test2' } });
    q.itemQuery.query({ request: { id: 'test3' } });

    const queries = queryClient.queryStore.getQueries(
        ItemQuery,
        (query) => !!query.variables.request?.id.includes('t')
    );
    expect(queries.length).toBe(3);

    const queries2 = queryClient.queryStore.getQueries(
        ItemQuery,
        (query) => !!query.variables.request?.id.includes('o')
    );
    expect(queries2.length).toBe(0);
});
```

```
 FAIL  tests/mstQuery.test.tsx > useQuery - reactive request
AssertionError: expected 1 to be 2 // Object.is equality
 ❯ tests/mstQuery.test.tsx:221:27
    219|     const queries = queryClient.queryStore.getQueries(ItemQuery, (query) => query);
    220| 
    221|     expect(queries.length).toBe(2);
       |                           ^
    222| 
    223|     configureMobx({ enforceActions: 'observed' });

  - Expected   "2"
  + Received   "1"

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  tests/mstQuery.test.tsx > findAll
 ❯ tests/mstQuery.test.tsx:479:27
AssertionError: expected 1 to be 3 // Object.is equality
    477|         (query) => !!query.variables.request?.id.includes('t')
    478|     );
    479|     expect(queries.length).toBe(3);
       |                           ^
    480| 
    481|     const queries2 = queryClient.queryStore.getQueries(

  - Expected   "3"
  + Received   "1"

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

Test Files  1 failed (1)
     Tests  2 failed | 36 passed (38)
  Start at  02:11:23
  Duration  1.15s (transform 357ms, setup 0ms, collect 223ms, tests 205ms)
  ```

**Comment by k-ode** [comment]
Thanks for raising this issue!

Currently, we don't cache data at the request level, but I do have long-term plans to implement such an API. The primary focus of this project has been on stability and bug fixes, which is why advanced caching has been put on the backburner for now.

As a workaround, you can use your own request cache at the network layer of your application.

```ts
const MessageQuery = createQuery('MessageQuery', {
    data: types.reference(MessageModel),
    request: types.model({ id: types.string }),
    endpoint({ request }) {
            const cachedResult = myTTLCache.get(request.id);
            if (cachedResult) {
               return cachedResult;
            }
            return fetch(...)
    },
});
```

**Comment by shovel-kun** [comment]
Thanks for the quick response,

What kind of caching is currently implemented as of `v3.4.0-canary.11`?

**Comment by k-ode** [comment]
There are two levels of caching:

**Using `staleTime` to control when a query should refetch**. 
This works similarly to how staleTime functions in react-query. When a query is re-mounted after its initial mount, it checks the staleTime to determine if the data is stale and whether it needs to refetch.

**Caching normalized models in their respective stores**. 
Normalized models are cached in their model stores and are only cleaned up when runGc is called.

If you want to ensure that a query runs only once per model, you can achieve this by fetching the cached model and setting the enabled prop of the query accordingly.

```tsx
const { itemStore } = useRootStore();

const cachedModel = itemStore.get(id);
const { data } = useQuery(q.itemQuery, {
    request: { id },
    initialData: cachedModel ? getSnapshot(cachedModel) : null,
    enabled: !cachedModel
});

```
